### ISSUE TITLE
Update zipcode-month panel in `derived`

### ISSUE THREAD

**Issue by diegogentilepassaro** [diegogentilepassaro/min_wage_rent#102]
Part of #85

[Goal]
Create a new updated zip code-month panel in `derived'

[Steps]
1. Use the new `base' level directories to create a zip code-month panel that allows us to estimate our baseline models. It needs to contain MWs, experiences MWs, rents, economic controls, and demographics. Keep all zip codes, regardless of whether they have rents data or not as that can be useful for assessing the representativeness of our data in terms of zip codes, but also in terms of what share of the MW changes we cover in our exercises.
2. Compare the number of zip codes, the number of zip code-month, and the number of treated zip codes with before the change in pipeline. Did we gain zip codes? 
3. Sanity check that we have MW data for the whole US in our 2010-2019 period. Create some maps. Compute how many events we use and how many there are at the state and local level. Also. what is the fraction of zip codes that experience a change in MW in the whole data and what is that number in our data that has rentals?

**Comment by diegogentilepassaro** [comment]
@gabrieleborg and @santiagohermo, I have created branch `issue102` to work in this task. I wanted to mention that I will use as a base panel the MW one, that should be for the whole US, and the I will merge the others to that one. Does that make sense to you?

**Comment by diegogentilepassaro** [comment]
A few initial notes @santiagohermo and @gabrieleborg : 
- Even if we start from the MW derived panel, we still need to join to our geo master to get other geographies than zip code. This is relevant for FEs and clustering.
- The `year_month` variable living in `../../../drive/derived_large/min_wage/zip_statutory_mw.csv` is not imported properly to STATA. We probably should make sure we use the proper options of fwrite for dates.
- The `zipcode` variable does not have leading zeros when zip codes start with 0. We should probably fix that by always saving them as a string.

**Comment by diegogentilepassaro** [comment]
@santiagohermo and @gabrieleborg, I started addressing [this comment](https://github.com/diegogentilepassaro/min_wage_rent/issues/102#issuecomment-805351420) and implementing the task. A few new issues that I encountered:

1. When merging the experienced MW panel to the baseline ZIPcode-year-month statutory MW panel we fail to match a lot of observations. It seems that this may get fixed when we have a working geo master, but flagging it in case you think of another reason that I should be aware of.
![Screen Shot 2021-03-24 at 6 33 11 PM](https://user-images.githubusercontent.com/20582494/112403536-f6718900-8ce4-11eb-8f89-a9a8c9fa3345.png)

2. When computing MW changes we still have a few negative ones. Flagging them here for the record so that we eventually make sure that they are legit.
![Screen Shot 2021-03-24 at 6 48 57 PM](https://user-images.githubusercontent.com/20582494/112403614-21f47380-8ce5-11eb-9a6e-4c02bd8ecb15.png)

**Comment by santiagohermo** [comment]
Thanks @diegogentilepassaro! On your points:

1. I agree, this should get fixed one we nail #101 
2. Most of the negative changes seem legit to me. The ZIP codes include belong to places listed [here](https://laborcenter.berkeley.edu/inventory-of-us-city-and-county-minimum-wage-ordinances/))

> Note that several minimum wage ordinances have been overturned, including Birmingham, AL; Miami Beach, FL; Louisville and Lexington, KY; Johnson, Lee, Linn, Polk, and Wapello Counties in Iowa; St. Louis, MO; and Kansas City, MO.

- For example, 33109, 33140 and 33141 are in Miami, were the state court overthrew a local ordinance in Feb 2019 (see [this news article](https://www.miamiherald.com/news/business/article225553350.html#:~:text=Miami%20Beach%20approved%20an%20ordinance,wage%20is%20%247.25%20an%20hour.))

- I am surprised about the ones in California (94608-94701). We have one MW changes in California on that date: Berkeley.
   
![image](https://user-images.githubusercontent.com/45404755/112409104-46554d80-8cef-11eb-82ea-08883427bacd.png)

This looks like a mistake in the raw data. Check, for example, page 4 [here](https://www.cityofberkeley.info/uploadedFiles/Clerk/Elections/2016/Berkeley%20Minimum%20Wage%20City%20-%20Full%20Text.pdf).

![image](https://user-images.githubusercontent.com/45404755/112409487-f88d1500-8cef-11eb-93fb-ee23f30e8bbe.png)

This was probably introduced because the "VZ" data had the last date of the panel with "current", and we never dropped that column when updating. I just fixed it, can you confirm it looks ok? Re-running `base/min_wage` should fix this.

Nice that you checked @diegogentilepassaro! One error less :)

**Comment by diegogentilepassaro** [comment]
@santiagohermo thanks for that! I re-run the code from `base/min_wage` and `derived/min_wage` and the negatives from California are now gone 👍 

New spotted issue: when merging `zip_demo_2010` coming from `base/demographics` we get new unmatched zip codes.
![Screen Shot 2021-03-25 at 1 30 09 PM](https://user-images.githubusercontent.com/20582494/112517067-47798f80-8d6e-11eb-98dc-d710519c61b3.png)

The ones with `_merge == 1` are fine because they are **all** just wrong matches coming from point 1. [in this comment](https://github.com/diegogentilepassaro/min_wage_rent/issues/102#issuecomment-806284611). However, for `_merge == 2` those are new unmatched observations. We have 1209 zip codes that have demographics and not a statutory MW assigned.  
![Screen Shot 2021-03-25 at 1 26 43 PM](https://user-images.githubusercontent.com/20582494/112517266-80196900-8d6e-11eb-9cce-7c71fff39fe6.png)

Should I expect this ones to also go away when we have the geo master fixed? 

FYI @gabrieleborg

**Comment by diegogentilepassaro** [comment]
@gabrieleborg I just realized about something a little weird in `base/demographics`. There is a script called `make_zippop_year.R` but it outputs a file called `acs_population_zipmonth.csv`. What should it be? In any case, at the base level let's agree to keep thing at the year and month and not year_month because dates are handled differently in R and STATA and matching year_months across datasets can be actually challenging! Could you please apply those changes?

Be aware that I added a line of code in `make_zippop_year.R` so that we output also a .dta file. Also, make sure to apply the changes in the branch issue_102 :D

**Comment by gabrieleborg** [comment]
ok!

**Comment by santiagohermo** [comment]
Thanks @diegogentilepassaro @gabrieleborg! Just to pitch in, maybe we can produce that file at the yearly level and keep it around. We may want to try using it in some specifications.

**Comment by diegogentilepassaro** [comment]
> ok @diegogentilepassaro !
> 
> The `acs_population_zipmonth.csv` contained the population control we wanted to include at a certain point. let's decide if we want to keep it.

Let's definitely keep it @gabrieleborg! I just didn't understood if it is at the yearly or monthly level. Which one is it? If it is monthly that is great :D

**Comment by santiagohermo** [comment]
If I remember correctly the data varies at the yearly level but is defined at the year month level.

**Comment by santiagohermo** [comment]
There are two branches related to this issue: `issue102` and `issue_102`. Per [here](https://github.com/diegogentilepassaro/min_wage_rent/commit/452bc19917afb0f1a752ae09f95d301bc183079b), it seems that `issue_102` is the one, so I'll delete the other one.

**Comment by santiagohermo** [comment]
Notes: Branch updated after #101 #103. Geo_master and Lodes OD matrices are updated here now.

I'll set the experienced MW to run.

**Comment by santiagohermo** [comment]
I rebuilt `derived/min_wage` @diegogentilepassaro @gabrieleborg. See data_manifest diff [here](https://github.com/diegogentilepassaro/min_wage_rent/commit/efec615f44a7d6f614b09c2fc6dc35cf9a3e0fb0#diff-6da9bd85d34a114ed142cfa01a02679c0e9d427d89936b474b0a3532c735609f)

Consider the zip code level data. In the statutory min wage dataset we gain like 400K observations, which is great. However, in the experienced MW we only gain 240 observations. Not what I was expecting! Also not sure if this is an issue or not (maybe we are missing zip codes that have no population or stuff like that). 

Maybe we can match these datasets and see what ZIP codes we are missing to understand what's going on?

**Comment by santiagohermo** [comment]
Another idea is that maybe we can leave a variable like "share_work_same_zipcode" in the dataset, so that we can later report how this looks like. (I think this was one of the comments we received in the last presentations)

**Comment by gabrieleborg** [comment]
Hey @santiagohermo @diegogentilepassaro ! 

Thanks for updating this :) 

> Consider the zip code level data. In the statutory min wage dataset we gain like 400K observations, which is great. However, in the experienced MW we only gain 240 observations. Not what I was expecting! Also not sure if this is an issue or not (maybe we are missing zip codes that have no population or stuff like that).

I believe the reason is in the new approach we use in `geo_master`. @santiagohermo updated branch #102 after #103 3 days ago. This branch was created 7 days ago by @diegogentilepassaro from #85. At that time, the latest commit to `base/geo_master/` is 40371d775f37947e38951a6c5abfe9ab95524300, originating from merging #99 into #85. if you look at `base/geo_master/code/build.do` in that commit, at line 57-82 you notice how we were still implementing the zcta-place-county approach, where for each combination we would only select 1 zipcode. If you take a quick look at [`raw/crosswalk/zip_to_zcta_2019.xlsx`](https://github.com/diegogentilepassaro/min_wage_rent/blob/master/raw/crosswalk/zip_to_zcta_2019.xlsx), and you filter ZCTA by keeping only those that have more than one ZIP CODE inside (i.e. you filter by `zip_join_type = Spatial Join to ZCTA`) you'll notice that there are approx 8000 zip codes: 
<img width="888" alt="Screen Shot 2021-03-29 at 3 19 18 PM" src="https://user-images.githubusercontent.com/30894427/112842837-41531e00-90a2-11eb-8b3b-efce02356d32.png">
This means that - given the 120 periods in our panel -  we were losing approximately 4000*120 = 480,000 observations. This to me looks like it could be why, since now we aren't sacrificing any zip code anymore, we gain roughly that amount. On the experience MW side, we weren't losing many in the first place since that is built on the LODES OD matrix, where we don't use the geo_master to get from block groups to zip codes (in fact, we use the tract_zip small crosswalk).

**Comment by gabrieleborg** [comment]
Hi @diegogentilepassaro ! 

Thanks for your updates, I will take a look and try to get a sense of why we have those mis-matches in the data!

**Comment by gabrieleborg** [comment]
Hi guys @diegogentilepassaro @santiagohermo !

Brown still hasn't restored access to Google File Stream, so I cannot access the repo yet. I thought that I could start posting a recap of what I found so far about the mis-matching issue, so we have a record of it even if there's still some non-matched zip codes that are unaccounted for (specifically, those present in the Experienced MW and not in the Statutory MW, `_merge==2`). 

I ran @diegogentilepassaro's `merge_all_data_sources.do` script and analyzed the merging of statutory and experienced MW datasets. 

- First, I took all (unique, hence removing the time dimension of the panel) non-matched zip codes and checked which were PO boxes. To do that I used [`raw/crosswalk/zip_to_zcta_2019.xlsx`](https://github.com/diegogentilepassaro/min_wage_rent/blob/master/raw/crosswalk/zip_to_zcta_2019.xlsx), which reports this piece of information. we get that: 

    | Zip code type | Count | 
    |---------------|-------|
    | non-matched zip codes, TOTAL | 6010 | 
    | non-matched zip codes, PO boxes | 4795 |
    | non-matched zip codes, physical | 1215 |  
    
    So, dropping the 4795 PO boxes is not a problem as they have no residents. We are left now with 1215 zip codes to account for. 

- Second, among these 1215 zip codes I started looking at those present in the statutory MW but not in the experienced MW. The latter data set is built using our "small geo-master", which is a tract-zip crosswalk. If they are not there it means that for some reason these zipcodes are either not present in the original `raw/crosswalk/TRACT_ZIP_122019.xlsx` file, or that are lost in the process of building the experienced MW using the LODES od matrix. I check both cases and report 1) how I did that, and 2) the numbers i find:
    - **zip codes not present in `TRACT_ZIP_122019.xlsx`**: To do this, I simply check which of the 1215 zip codes are not present in the raw `TRACT_ZIP_122019.xlsx` file. The number is 44. 
    - **zip codes lost in the Exp MW building process**: To do this, I went through all steps we use to buld Exp MW. First, we bulid the od matrix at the zipcode level. This is obtained by aggregating the original, block-group level matrix using our small geo-master, that in turn is bascially a clean version of `TRACT_ZIP_122019.xlsx` crosswalk. Here I discovered that the missing zip codes are those whose underlying tracts have a residential share of 0. Since we use that measure as our preferred weight, what happens is that when we try to aggregate workers flows from tracts falling within a zip code's boundary, since they all have zero residents the associated zip codes disappear in the aggregating R function `base/lodes/code/make_odmatrix.R`: 
        <img width="752" alt="Screen Shot 2021-04-01 at 11 06 57 AM" src="https://user-images.githubusercontent.com/30894427/113271021-629e4f00-92da-11eb-83fc-219ef85d0c25.png">
     What this means is that, even if that tract is sending some workers in some other tract, they are sending them from a part of the tract that does not fall within the boundaries of the zip codes under analysis. So we are not "losing" that flow, but simply it will be accounted in another zip codes where that tract will have a residential share larger than zero. NOTICE that I checked this manually, as I couldn't go over all of them. But I randomly checked something like 20 zipcodes and they all have this feature (see exampe in the pictures):  
    <img width="238" alt="Screen Shot 2021-03-30 at 3 42 20 PM" src="https://user-images.githubusercontent.com/30894427/113271707-19023400-92db-11eb-9fa1-8b5fec4e9a92.png">
    
    <img width="327" alt="Screen Shot 2021-03-30 at 3 40 45 PM" src="https://user-images.githubusercontent.com/30894427/113271745-21f30580-92db-11eb-8583-550d83b30ca8.png">
    

    - **Recap**: 
    
     | Zip code type | Count | 
    |---------------|-------|
    | non-matched zip codes, physical, `_merge==1`,  TOTAL | 710 |
    | non-matched zip codes, physical, `_merge==1`,  not in TRACT_ZIP_122019.xlsx | 44 |
    | non-matched zip codes, physical, `_merge==1`, res_ratio = 0 | 666 | 
 
    Dropping the 666 non-matched zip codes is not a problem because workers are accounted somewhere else, but we may want to quickly check whether there is some overlapping between the 44 missing zip codes in the crosswalk and the zillow zip codes. If not, we can drop all non-matched. 

Lastly, as I previously said there are still 505 zip codes that need investigations (1215 - 710 = 505). These are non-matched, physical zip codes where `_merge==2`, i.e. that are present in the experienced MW but not in the statutory.

**Comment by diegogentilepassaro** [comment]
Hi @gabrieleborg! Thank you very much for the detailed exploration of the weird zip codes :D Sorry that I just get the time to answer but I was making a lot of changes to the pipeline to use the geo master everywhere and to make things work in a very straightforward way. Note that I also made changes to the geo master and I also fixed a couple of mistakes along the way too! The basic structure that I am following now is to build 4 panels:
1. A zipcode panel.
2. A zipcode-month panel.
3. A county panel.
4. A county-month panel.

At this stage I think that we want to check that the numbers and the merges make sense and after that I will generate a new folder in `derived` to add the coding for the balancing and that kind of operations that we were applying before. Could you please take a look and pseudoPR that my changes make sense to you?

Some important comments:
- I made quite some assertions to make sure that we keep all of our zip codes that come from with rental data. Let me know if they make sense to you and if you think about other assertions. 
- I left comments along the script signaling future steps and things that ideally we would like to do. Once you think about them, let's discuss!
- Let's see if after the changes if we still have the issues described in [your comment](https://github.com/diegogentilepassaro/min_wage_rent/issues/102#issuecomment-811778123) and if they are as big. The way I was assessing that was by checking the merges at the zipcode-month build. 
- We should probably apply some changes so that we save base datasets such that zip code is always a string and year and month variables are always integers to ensure compatibility of merges. Let me know if you agree :D

Tomorrow in the meeting we can discuss further!

**Comment by santiagohermo** [comment]
Thanks for that great work @diegogentilepassaro! Maybe it's easier if you go to PR and @gabrieleborg reviews the code there? We can always open a follow-up issue to implement next steps :)

**Comment by santiagohermo** [comment]
Thanks [for that](https://github.com/diegogentilepassaro/min_wage_rent/issues/102#issuecomment-811778123) @gabrieleborg! Really useful. Given your exploration, do you think that we should set experienced = statutory in the case when the experienced is missing?

**Comment by gabrieleborg** [comment]
Thanks @diegogentilepassaro for the updates :) 

I briefly reviewed your edits, and they look fine to me! Few comments/questions: 

1. I noticed that in `base/geo_master` we now output 3 datasets: a zipcode place county-level geo master, a tract zipcode crosswalk and a new zipcode level geo master. This last one, `zip_county_place_usps_master` is basically the geo master when choosing - for each zip code -  the zipcode place county combination with the most houses. Is that correct? When were you thinking about using this? I am fine with creating it, I am just curious because I didn't spot where you later used it!
2.  You created a `zipcode`, `zipcode_month`, `county`, `county_month` subfolders in derived. So far the only think I noticed is that in the former you use the zipcode-level geo master, but I am still not sure about the goal for that. What is the main difference between the normal, and the -month version of the panels (also in terms of practical application downstream)?
3. Curious to see what happens with the merge issue! keep us updated !

**Comment by gabrieleborg** [comment]
Hey @santiagohermo !

> Given your exploration, do you think that we should set experienced = statutory in the case when the experienced is missing?

I think that yes, we should do that since It makes sense from a logical point of view. For those zip codes where we don't have residents, the exp MW technically is the statutory one. But I would be a bit more careful about those missing 44 zip codes that are not present in the original raw crosswalk. For those, we don't really know as they may actually have some residents. Also, Brown still hasn't restored Google File Stream, so It's difficult for me to finish investigating the missing zip codes :(  Let's discuss this tonight
